pkg/kv/kvserver/store.go

```diff
@@ -1893,7 +1893,12 @@ func (s *Store) Start(ctx context.Context, stopper *stop.Stopper) error {
 		// constraint (*Replica).raftMu < (*Store).mu. See the comment on
 		// (Store).mu.
 		s.mu.Lock()
-		err = s.addReplicaInternalLocked(rep)
+		// TODO(pavelkalinnikov): hide these in Store's replica create functions.
+		// TODO(pavelkalinnikov): addReplicaToRangeMapLocked needs a locked repl.mu.
```

Comment:
(pre-existing)

Reply:
However, this might be ok since it's the "initialization" path. But it's not super-obvious. `addReplicaToRangeMapLocked` inserts the replica to the map **before** it accesses `repl.mu.quiescent`.

Reply:
Huh, I looked at `addReplicaToRangeMapLocked` and that it needs to look at the quiescing status seems fishy. Concretely, since replicas are always added to this map right when they're created (and are necessarily uninitialized - I'm ignoring the weirdo`(*Store).Start` here), how would they become unquiesced? An uninited replica physically *cannot* unquiesce:

https://github.com/cockroachdb/cockroach/blob/d40095138d8eff517f9126b71cbc7e9335ac0c30/pkg/kv/kvserver/replica_raft_quiesce.go#L95-L122

So the insertion into the quiescent there should be unconditional (and wouldn't need `r.mu` at all except maybe to assert).

For splits, I can only guess that the comment refers to

https://github.com/cockroachdb/cockroach/blob/38ad1f3cf9023329e84d1c0e961afc21bcf77c11/pkg/kv/kvserver/store_split.go#L355

but we already acquired the split lock, which already inserted the replica, so the second call [here](https://github.com/cockroachdb/cockroach/blob/0703807d018fb1c0f5352e8d83315b463c90d1e9/pkg/kv/kvserver/store_create_replica.go#L342) still has the replica in the map, and so there is no need to mess with the quiesced status. There's no such call on merges, so I'm hard-pressed to say what this is about. Possibly the comment predates the split/merge triggers in their entirety.

Reply:
> So the insertion into the quiescent there should be unconditional (and wouldn't need `r.mu` at all except maybe to assert).

I think this branch should rather be removed, it never happens (it inserts to ***un***quiesced).

Unquiescence can only happen on a `Replica` that's linked to `Store`, in reaction to some events/messages (right?). On top of that, as you say, unquiescence can only happen for initialized replicas. At the end state of this PR, `addToReplicasByRangeIDLocked` is called only for uninitialized replicas that aren't linked to `Store` and can't have got any messages.

This check was added in #27518, was it making more sense back then?

Reply:
Added another commit which removes this check. See the commit message with explanation.

Reply:
Thanks for digging up the PR that introduced this. This adds further weight that the check isn't useful: the PR's goal was to *avoid* adding uninited replicas as unquiesced. I assume the check was necessary because the call structure around replica lifecycles was even more circuitous then. :-)